dhttp/server.go

```diff
@@ -0,0 +1,371 @@
+//    methods.
+//  - When shutting down, it properly blocks when waiting for the workers of hijacked connections.
+//  - If you use dlog, you don't have to manually configure the logging for the server to do the
+//    right thing.
+//
+// Breaking changes from http.Server to ServerConfig that will stop your old code from compiling:
+//
+//  - Obviously, the name is different.
+//  - The "Addr" member field is removed; it is replaced by an "addr" argument to the
+//    "ListenAndServe(TLS)?" methods.
+//  - The "BaseContext" member field is removed; it is replaced by a "ctx" argument to the
+//    "(ListenAnd)?Serve(TLS?)" methods.
+//  - The "RegisterOnShutdown" is removed; it is replaced by an "OnShutdown" member field.
+//  - The "SetKeepAlivesEnabled", "Shutdown", and "Close" methods are removed; they are conceptually
+//    replaced by using Context cancellation for lifecycle management.  Use dcontext soft
+//    cancellation for the graceful shutdown that "Shutdown" allowed.
+//
+// Breaking changes from http.Server to ServerConfig that will maybe make your old code incorrect:
+//
+//  - The semantics of the "TLSNextProto" member field are slightly different.
+//  - The semantics of the "Error" member field are slightly different.
+//  - The structure is deep-copied by each of the "(ListenAnd)?Serve(TLS?)" methods; mutating the
+//    config structure while a server is running will not affect the running server.
+//  - HTTP/2 support (both "h2" and "h2c") is built-in, so if your code configures HTTP/2 manually,
+//    you're going to need to set "DisableHTTP2: true" to stop ServerConfig from stomping over your
+//    code's work.
+//
+// Arguably-breaking changes from http.Server that to ServerConfig that I'd say are bugfixes, but
+// could conceivably[1] make someone's old code incorrect:
+//
+//  - *http.Server.ServeTLS won't close the Listener if .ServeTLS returns early during setup due to
+//    having been passed invalid cert or key files; ServerConfig.ServeTLS will always close the
+//    Listener before returning; matching the "Serve" method.
+//
+// The reason for creating a new type and having breaking changes (rather than writing a few utility
+// functions that take an *http.Server as an argument) is that it became increasingly clear that the
+// lifecycle of a running server is tied to the lifecycle of the *http.Server object, while we've
+// grown a standard "Context" lifecycle system that wants to tie the lifecycle to the function call,
+// Rather than to the object.  This divorce between the running-server lifecycle and the object is
+// embodied in the name of the type; when the lifecycle of the object was the lifecycle of the
+// server, the object *was* the server, so it was named *http.Server; now the lifecycle is divorced
+// from that, and the object is just configuration for the server, so it is named *ServerConfig.
+//
```

Comment:
I think this can be usefully split up: "This divorce between [lifecycles] is embodied in the name of the type: when the lifecycle was [tied to the object]... it was named *http.Server. Now the lifecyle is divorced from that... so [the object] is named *ServerConfig."